File: tools/e2e-cli/src/tests/publish.ts
import assert from 'assert';

import { TestContext, TestDefinition } from '../types';

async function testPublish(ctx: TestContext): Promise<void> {
  const packages = [
    `verdaccio-memory-${ctx.runId}`,
    `verdaccio-${ctx.runId}`,
    `@verdaccio/foo-${ctx.runId}`,
    `@verdaccio/some-foo-${ctx.runId}`,
  ];

  const type = ctx.adapter.type;

  for (const pkgName of packages) {
    const { tempFolder } = await ctx.adapter.prepareProject(
      pkgName,
      '1.0.0',
      ctx.registryUrl,
      ctx.port,
      ctx.token
    );

    // yarn modern requires install before publish to generate lockfile
    if (type === 'yarn-modern') {
      await ctx.adapter.exec({ cwd: tempFolder }, 'install');
    }

    const resp = await ctx.adapter.exec(
      { cwd: tempFolder },
      'publish',
      '--json',
      ...ctx.adapter.registryArg(ctx.registryUrl)
    );

    if (type === 'yarn-modern') {
      assert.ok(
        resp.stdout.match(/Package archive published/),
        `Expected "Package archive published" for ${pkgName} but got "${resp.stdout}"`
      );
    } else if (type === 'yarn-classic') {
      // yarn classic --json outputs NDJSON — just verify no error exit (exit code 0 is enough)
      assert.ok(resp.stdout.length > 0, `Expected publish output for ${pkgName}`);
    } else {
      // npm / pnpm
      try {
        const parsedBody = JSON.parse(resp.stdout);
        assert.strictEqual(parsedBody.name, pkgName, `Expected package name "${pkgName}" but got "${parsedBody.name}"`);
        assert.ok(parsedBody.files, `Expected files to be defined for ${pkgName}`);
      } catch {
        // pnpm v11+ native publish doesn't output JSON — exit code 0 is sufficient
        assert.ok(true, `Publish succeeded for ${pkgName}`);
      }
    }
  }
}

export const publishTest: TestDefinition = {
  name: 'publish',
  requires: ['publish'],
  run: testPublish,
};

File: tools/e2e-cli/src/e2e_tests/publish.rs
use super::prelude::*;

pub(crate) fn test_publish(ctx: &mut TestContext<'_>) -> Result<()> {
    let packages = vec![
        format!("verdaccio-memory-{}", ctx.run_id),
        format!("verdaccio-{}", ctx.run_id),
        format!("@verdaccio/foo-{}", ctx.run_id),
        format!("@verdaccio/some-foo-{}", ctx.run_id),
    ];
    for pkg_name in packages {
        let temp = prepare_project(ctx, &pkg_name, "1.0.0", BTreeMap::new(), BTreeMap::new())?;
        if ctx.adapter.kind == AdapterType::YarnModern {
            exec_adapter(ctx, Some(&temp), vec!["install".into()])?;
        }
        let mut args = vec!["publish".into(), "--json".into()];
        args.extend(registry_arg(&ctx.adapter, &ctx.registry_url));
        let resp = exec_adapter(ctx, Some(&temp), args)?;
        match ctx.adapter.kind {
            AdapterType::YarnModern => assert_contains(&resp.stdout, "Package archive published")?,
            AdapterType::YarnClassic => {
                assert_true(!resp.stdout.is_empty(), "Expected publish output")?
            }
            AdapterType::Bun => {}
            _ => {
                if let Ok(parsed) = serde_json::from_str::<Value>(&resp.stdout) {
                    let expected_id = format!("{pkg_name}@1.0.0");
                    let published_name = parsed.get("name").and_then(Value::as_str);
                    let published_id = parsed.get("id").and_then(Value::as_str);
                    if published_name.is_some() || published_id.is_some() {
                        assert_true(
                            published_name == Some(pkg_name.as_str())
                                || published_id == Some(expected_id.as_str()),
                            &format!(
                                "Expected package name \"{pkg_name}\" or id \"{expected_id}\""
                            ),
                        )?;
                    }
                }
            }
        }
        assert_published(ctx, &pkg_name)?;
    }
    Ok(())
}

fn assert_published(ctx: &TestContext<'_>, pkg_name: &str) -> Result<()> {
    let encoded = pkg_name.replace('/', "%2F");
    let response = Client::new()
        .get(format!(
            "{}/{}",
            ctx.registry_url.trim_end_matches('/'),
            encoded
        ))
        .bearer_auth(&ctx.token)
        .send()
        .with_context(|| format!("failed to fetch published package {pkg_name}"))?;
    let status = response.status();
    let body: Value = response.json().unwrap_or(Value::Null);
    assert_true(
        status.is_success(),
        &format!("Expected published package {pkg_name}, got {status} {body}"),
    )?;
    assert_eq_json(
        body.get("name"),
        pkg_name,
        "Expected published registry package name",
    )
}
